### DISCUSSION TITLE
error in makefile commands

### DISCUSSION THREAD

**Discussion by khushbusandhu** [Cyfrin/foundry-full-course-cu#649] [General]
on running 
```
make deploy ARGS="--network sepolia"
```

i got this error 
```
Error: 
Could not canonicalize the target path

Context:
- No such file or directory (os error 2)
make: *** [Makefile:43: deploy] Error 1
```

**Comment by theirrationalone** [comment]
would you like to share your makefile and its naming??

**Reply by khushbusandhu** [reply]
here it is 
```
-include .env

.PHONY: all test clean deploy fund help install snapshot format anvil 

DEFAULT_ANVIL_KEY := 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80

help:
	@echo "Usage:"
	@echo "  make deploy [ARGS=...]\n    example: make deploy ARGS=\"--network sepolia\""
	@echo ""
	@echo "  make fund [ARGS=...]\n    example: make deploy ARGS=\"--network sepolia\""

all: clean remove install update build

# Clean the repo
clean  :; forge clean

# Remove modules
remove :; rm -rf .gitmodules && rm -rf .git/modules/* && rm -rf lib && touch .gitmodules && git add . && git commit -m "modules"

install :; forge install Cyfrin/foundry-devops@0.0.11 --no-commit && forge install smartcontractkit/chainlink-brownie-contracts@0.6.1 --no-commit && forge install foundry-rs/forge-std@v1.5.3 --no-commit && forge install transmissions11/solmate@v6 --no-commit

# Update Dependencies
update:; forge update

build:; forge build

test :; forge test 

snapshot :; forge snapshot

format :; forge fmt

anvil :; anvil -m 'test test test test test test test test test test test junk' --steps-tracing --block-time 1

NETWORK_ARGS := --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast

ifeq ($(findstring --network sepolia,$(ARGS)),--network sepolia)
	NETWORK_ARGS := --rpc-url $(SEPOLIA_RPC_URL) --private-key $(PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
endif

deploy:
	@forge script script/DeployRaffle.s.sol:DeployRaffle $(NETWORK_ARGS)

createSubscription:
	@forge script script/Interactions.s.sol:CreateSubscription $(NETWORK_ARGS)

addConsumer:
	@forge script script/Interactions.s.sol:AddConsumer $(NETWORK_ARGS)

fundSubscription:
	@forge script script/Interactions.s.sol:FundSubscription $(NETWORK_ARGS)
```

**Reply by khushbusandhu** [reply]
on running 
```
forge script script/Deployraffle.s.sol --fork-url $SEPOLIA_RPC_URL
```
 everything worked but there is no verification part that occurred why so?

**Reply by theirrationalone** [reply]
you're missing --verify --etherscan-api-key $ETHERSCAN_API_KEY

**Reply by khushbusandhu** [reply]
error is there 
```

(code: -32000, message: Must be authenticated!, data: None)
```

**Reply by theirrationalone** [reply]
in .env file, make sure to put data like this

ETHERSCAN_API_KEY=YOUR_KEY // without white space before after =

**Reply by khushbusandhu** [reply]
yes it is same you said 
but is the API KEY TOKEN is our API key

**Reply by theirrationalone** [reply]
Yes of course!

**Reply by khushbusandhu** [reply]
then what's the problem with this😢

**Reply by theirrationalone** [reply]
delete that image link above first.

**Reply by khushbusandhu** [reply]
done

**Reply by theirrationalone** [reply]
see this, are you doing this exactly...

![ss etherscan](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/ae510ceb-dc89-4afd-97e6-aa9fa612d1db)

**Reply by theirrationalone** [reply]
but for sepolia network

**Reply by theirrationalone** [reply]
or try to restart, as shikhar says

**Reply by theirrationalone** [reply]
Heyy, You might omitting to append that command with `--broadcast` ??

**Reply by theirrationalone** [reply]
Hope that's now resolved (verification)...

But I also didn't get that `Could not canonicalize the target path` thing in Makefile. However, I asked about it from `ChatGPT`🚀 but it's saying something about Relative/Absolute Path thing..... I checked your `Makefile` but everything seems Okay in that. Sorry for it, One Experiment we can do here...

try to prepend all paths with `./` like this...

```makefile
deploy:
	@forge script ./script/DeployRaffle.s.sol:DeployRaffle $(NETWORK_ARGS)
```

or you can research yourself until you got the answer either from here or from any other resources.....

**Reply by khushbusandhu** [reply]
I have run the command with --broadcast also but it gave the error. 
But I haven't tried to find its solution right now because I have class tomorrow at 8am and now I am going to sleep so I will do the rest tomorrow after classes.

**Reply by theirrationalone** [reply]
Actually you're trying to verify contract onto the Forked Network, review your command there's `--fork-url`.

and verifying a contract on an imaginary network is impossible. 

fork network `forknet` is just a simulated real network `mainnet` / test network `testnet` it's for testing how our contract will perform on `mainnet`|`testnet` / real|test network. So it doesn't make sense to verify a contract on an imaginary network which has zero support of etherscan.

**Reply by theirrationalone** [reply]
I was thinking about this error in my mind `
(code: -32000, message: Must be authenticated!, data: None)`.... And i figured out what if we were verifying our contract on an imaginary network ooohh heck this could also be a problem and i also asked from chatgpt however in result chatgpt also suggested me similar kind of issues. And i came back to this discussion and checked your rpc(remote procedure call) 's reference... After that i caught that `--fork-url`. 

You should use `--rpc-url`.

**Reply by khushbusandhu** [reply]
Now the usual error is there
'''
error sending request for url (https://eth-sepolia.g.alchemy.com/v2/ooAbJoM0xj2HSJby6aYZi82Eetcz0Q3_): error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution

Context:
- Error #0: error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution
- Error #1: dns error: failed to lookup address information: Temporary failure in name resolution
- Error #2: failed to lookup address information: Temporary failure in name resolution
- ''''

**Reply by theirrationalone** [reply]
Have you Resolved that on your own??

**Reply by khushbusandhu** [reply]
yes it is resolved but new error occurred and now I am trying searching about that . here's the error
```

Script ran successfully.

== Return ==
0: contract Raffle 0x90193C961A926261B756D1E5bb255e67ff9498A1
1: contract HelperConfig 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f

== Logs ==
  Adding Consumer contract : 0x90193C961A926261B756D1E5bb255e67ff9498A1
  using vrfCoordinator: 0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625
  On ChainId: 11155111

## Setting up (1) EVMs.

==========================

Chain 11155111

Estimated gas price: 45.855685294 gwei

Estimated total gas used for script: 1997548

Estimated amount required: 0.091598932447659112 ETH

==========================

###
Finding wallets for all the necessary addresses...

Transactions saved to: /home/khushbu/foundry-smart-contract-lottery-f23/broadcast/Deployraffle.s.sol/11155111/run-latest.json

Sensitive values saved to: /home/khushbu/foundry-smart-contract-lottery-f23/cache/Deployraffle.s.sol/11155111/run-latest.json

Error: 

You seem to be using Foundry's default sender. Be sure to set your own --sender.
No associated wallet for addresses: {0x1804c8ab1f12e6bbf3894d4083f33e07309d1f38}. Unlocked wallets: [0x00af730768898787e368e5df09e6fd2a0fb5ac27]
```

**Reply by khushbusandhu** [reply]
hey @theirrationalone  all of these error are resolved now and look what i got 
```
 Contract Source Code Verified (Exact Match)
```

**Reply by theirrationalone** [reply]
Congrats! 🎉🎉 You didi all that on your own 🚀 🚀🚀

**Reply by khushbusandhu** [reply]
but performUpkeep is not running on the automation chainlink.

**Reply by theirrationalone** [reply]
Make sure you provided correct subscription id and performed `enterRaffle` through interactions on on sepolia `testnet`, Next after doing this checkout vrf & automation website to verify all the effects.... 

You can check contract's transactions by going to your deployed contract on sepolia etherscan... On there click on transactions tab there you can see `enterRaffle` entry... 

And by clicking on internal txs. Tab, on there you can see your events emitted... Where you can verify checkupkeep, performupkeep, fulfillrandomwords, txs... After waiting 30 seconds Or according to your interval...  Count From `enterRaffle` transaction performed.

**Reply by khushbusandhu** [reply]
i have followed the step done by Patrick
the error I realised is that while putting the contract address for registering new upkeep it shows this 
```
Unable to verify if this is an Automation compatible contract.
```
this is the only difference occurring from all the steps that Patrick did

**Reply by theirrationalone** [reply]
You might messed up with dependent functions visibility, Plzz verify that you're giving right visibility to functions.... Internal for `fulfillRandomwords` and so on...

**Reply by theirrationalone** [reply]
I've checked out your repo `foundry-lottery` .... 

Actually you forgot to import & inherit library which supports automation... 

This one:
```solidity
import {AutomationCompatibleInterface} from "@chainlink/contracts/src/v0.8/interfaces/AutomationCompatibleInterface.sol";
```

And don't forget to inherit it. 

Like... 

```solidity
contract Raffle is AutomationCompatibleInterface {
    // statements... 
```

**Reply by khushbusandhu** [reply]
okay got it 
one more silly mistake in count😢

**Reply by theirrationalone** [reply]
Will concrete concepts. 😅

**Reply by khushbusandhu** [reply]
These type of  lines motivates me alot to do my work and not think about result because that will be going to be something good as a by product of all this
